Add upload completion processor for VideoPress block

SiobhyB opened the pull request:
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5571

### Details

An media upload completion processor has been added for the new VideoPress block. The processor follows similar patterns to other media processors, particularly the processor for [the audio block](https://github.com/wordpress-mobile/WordPress-Android/blob/387847010b8884939d2e10f3ceb7f9a4b852114d/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessor.kt), given that was the most recently added.

If a user navigates away from the editor while the VideoPress block is uploading, the processor will complete the upload and populate it with the expected content. In order to achieve, it's necessary to build a VideoPress URL based on the block's various settings. You can see how this is achieved in Jetpack [here](https://github.com/Automattic/jetpack/blob/b1b826ab38690c5fad18789301ac81297a458878/projects/packages/videopress/src/client/lib/url/index.ts#L19-L67), and a similar pattern has been followed in this PR.

The iOS equivalent of this PR can be found at https://github.com/wordpress-mobile/WordPress-iOS/pull/20380

### Testing

1. Add a VideoPress block to an existing or new post.
2. Select the **choose from device** option and upload a video.
3. While there’s an ongoing upload, close the post with publishing changes.
4. Verify that you see the upload progress in the post summary.
5. Re-open the post and confirm the upload has completed.
6. Preview the post and observe that the video can be played.

### Default markup examples

**Example default markup before upload:**

```
<!-- wp:videopress/video {"id":"1234" -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"></figure>
<!-- /wp:videopress/video -->
```

**Example default markup after upload:**

```
<!-- wp:videopress/video {"id":"1234","guid":"AbCdE" -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">https://videopress.com/v/AbCdE?resizeToParent=true&amp;cover=true&amp;preloadContent=metadata&amp;useAverageColor=true</div></figure>
<!-- /wp:videopress/video -->
```

### Markup with attributes examples

**Example markup before upload:**

```
<!-- wp:videopress/video {"autoplay":true,"controls":false,"description":"","id":${localMediaId},"loop":true,"muted":true,"playsinline":true,"poster":"https://test.files.wordpress.com/2022/02/265-5000x5000-1.jpeg","preload":"none","seekbarColor":"#abb8c3","seekbarLoadingColor":"#cf2e2e","seekbarPlayedColor":"#9b51e0","title":"Demo title","useAverageColor":false} -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"></figure>
<!-- /wp:videopress/video -->
```

**Example markup after upload:**

```
<!-- wp:videopress/video {"autoplay":true,"controls":false,"description":"","id":${remoteMediaId},"loop":true,"muted":true,"playsinline":true,"poster":"https://test.files.wordpress.com/2022/02/265-5000x5000-1.jpeg","preload":"none","seekbarColor":"#abb8c3","seekbarLoadingColor":"#cf2e2e","seekbarPlayedColor":"#9b51e0","title":"Demo title","useAverageColor":false,"guid":"${videoPressGuid}"} -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">https://videopress.com/v/AbCdE?resizeToParent=true&amp;cover=true&amp;autoPlay=true&amp;controls=false&amp;loop=true&amp;muted=true&amp;persistVolume=false&amp;playsinline=true&amp;posterUrl=https%3A%2F%2Ftest.files.wordpress.com%2F2022%2F02%2F265-5000x5000-1.jpeg&amp;preloadContent=none&amp;sbc=%23abb8c3&amp;sbpc=%239b51e0&amp;sblc=%23cf2e2e</div></figure>
<!-- /wp:videopress/video -->
```

## Regression Notes
1. Potential unintended areas of impact

This change might impact other upload processors, with the video processor being a specific concern due to the possibility of the similar block names being matched.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I manually tested uploading media to both the video and VideoPress blocks. I also relied on the existing automated tests for each of the processors to flag any issues and added new tests for the VideoPress block.

3. What automated tests I added (or what prevented me from doing so)

* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/91cc0bdd7f5352bb0c04e51ef38d4ecd3bdbb843: Test to check the video block processor doesn't change anything in the VideoPress block.
* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/ec71ddbe594d64720f1dc371367204b9abba4af6: Test to check the VideoPress block processor doesn't change anything in the video block.
* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/f36eb72bd561f93fed04a66340a0c9ae3b2b8dd0: Test to ensure processor correctly converts VideoPress block with default attributes. I followed the same pattern set out [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/2ddbacb28a54ef608e1250256c4b05381d3662d4/WordPress/WordPressTest/GutenbergVideoPressUploadProcessorTests.swift#L7-L30) for iOS. 
* https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/4083c4d233659ae46c818d890cb24fe8cc31625b: Test to ensure processor correctly converts VideoPress block with non-default attributes. I followed the same pattern set out [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/2ddbacb28a54ef608e1250256c4b05381d3662d4/WordPress/WordPressTest/GutenbergVideoPressUploadProcessorTests.swift#L32-L55) for iOS.

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
feat: Add skeleton code VideoPress media processor

The skeleton code for the VideoPress block's media processor has been added with this commit. Its functionality will be fleshed out in future commits. For now, it follows the same pattern as other block processors.

The audio block was followed the most closely, given it was the newest processor at the time of writing: https://github.com/wordpress-mobile/WordPress-Android/blob/d609fa52b6c600da9c23f37f60053bbb79bb811b/WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessor.kt

SiobhyB committed:
feat: Add GUID to block's attributes

SiobhyB committed:
feat: Use attributes to define/update settings

The block's settings may differ from the defaults. We'll need to detect this in order to populate the block's content. As such, we'll define a class and use the information we get from the block's attributes to update the settings.

SiobhyB committed:
feat: Populate block's content, including URL

SiobhyB committed:
refactor: Split getVideoPressURL to smaller funcs

The code previously failed linting checks with the following error:

> The function getVideoPressURL appears to be too complex (23). Defined complexity threshold for methods is set to '15' [ComplexMethod]

As such, this commit refactors that function and seeks to reduce its complexity by splitting it into smaller, more readable functions.

SiobhyB committed:
refactor: Further simplify logic for building URL

Following the last commit, the linting checks still indicated that the logic for building the VideoPress URL was too complex. As such, this commit further simplifies the functionality by splitting each individual setting into its own function.

This now matches the approach taken in the iOS PR here: https://github.com/wordpress-mobile/WordPress-iOS/pull/20380/

SiobhyB committed:
test: video processor doesn't impact VideoPress

SiobhyB committed:
test: VP processor doesn't impact video block

In this commit, a "VideoPressBlockProcessorTest.kt" file has been created, with its first test. The test confirms that the video block remains unchanged when the VideoPress processor runs.

SiobhyB committed:
fix: Enable unit tests to mock URL

The Android-specific Uri class for encoding query parameters doesn't work with unit tests.

As such, it's been replaced with the java.net.URLEncoder class with this commit.

As the URLEncoder class isn't supported on versions of Android lower than 13 (and the app still officially supports versions down to Android 7), conditional logic is used to incorporate both.

SiobhyB committed:
test: process default VideoPress block

SiobhyB committed:
fix: Add missing setting

The 'playsinline' setting was not being correctly added to the VideoPress URL when set, due to it being missing from the class. This commit fixes that mistake.

SiobhyB committed:
test: process VP block with non-default settings

SiobhyB committed:
refactor: Signify code that requests SDK S

SiobhyB committed:
refactor: Follow lint recs to set funcs to private

SiobhyB committed:
fix: Ensure encoded quotes are removed from URL

SiobhyB committed:
refactor: Move encode func to wrapper for testing

With this commit, a wrapper has been added for the Uri class, so that we can easily reference the encode functionality in our unit tests. We also discard the usage of URLEncoder.encode for devices with API 33 or above, for simplicity. Uri.encode will work with all APIs.

SiobhyB committed:
test: Mock URL encoding

SiobhyB committed:
refactor: Remove unused imports

SiobhyB committed:
Merge branch 'trunk' into gutenberg/videopress-completion-processor

SiobhyB commented on WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/UriWrapper.kt:
```diff
@@ -0,0 +1,18 @@
+ * 
+ * We need to use the Uri.encode function in order to encode a URL for the VideoPress block.
+ * However, As Uri.encode is part an Android class, it cannot run in unit tests by default.
+ * Note, java.net.URLEncoder is not currently a suitable alternative as it require API 33 and above (our minimum is 24 at the time of writing)
```
Another alternative I explored to this approach included "forcing" the unit tests to use an API above 33 then conditionally only using `Uri.encode` for older APIs, while using `java.net.URLEncoder` for newer ones. However, I couldn't find a straightforward way to achieve this. I looked into using a framework like [Robolectric](https://robolectric.org/) but found discussions in which we'd previously decided to remove that framework in https://github.com/wordpress-mobile/WordPress-Android/pull/16170.

I ended up only using `Uri.encode` in the main processor, as this will work with all APIs that we support. Through the use of this wrapper, I was then able to use `java.net.URLEncoder` directly in the unit tests. It's my understanding that these functions work in pretty much identical ways, so the results from the unit tests will remain accurate.

I'd definitely be curious if others have other ideas for an approach that would work across both the main codebase and the unit tests.

SiobhyB committed:
refactor: Reduce line width to accommodate linter

SiobhyB committed:
fix: Prevent local URL being stored in JSON

SiobhyB commented:
~@mkevins, 👋 , I saw you've had a lot of involvement in the other media processors, would you be able to take a look over this one? Your native-expertise would be really appreciated~ 🙇‍♀️  I just realised you're on leave, please ignore this ping! :)

fluiddot commented on WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/UriWrapper.kt:
I noticed we already have a wrapper for the same issue with unit tests [`UriWrapper`](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/util/UriWrapper.kt). I'm wondering if we could reuse it and incorporate the `encode` function, WDYT?

SiobhyB replied:
Oh, I hadn't spotted that existing wrapper! 🤦‍♀️ Thank you for pointing that out. I agree with re-using it, and am happy the existence of an existing wrapper supports that this is an okay approach to go down (I'd had [some uncertainty](https://github.com/wordpress-mobile/WordPress-Android/pull/18187#discussion_r1152250160)).

fluiddot replied:
Yeah, I was kind of shocked that `Uri.encode` can't be used in unit tests, but after checking the code and googling a bit, seems this is a known issue. Good call 🙂 !

SiobhyB replied:
~Addressed this in https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/594420cd3f0f268a6bed95c11b96cec38c352d6d.~ Edited: This needs more work to work with unit tests.

SiobhyB replied:
Addressed the changes in https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/db6fb39fffd6f5af79c4845535ea7fb972733697. Note, I ended up creating a separate `UriEncoder` class within the file as this seemed the most straightforward approach for both allowing it to be called with no Uri argument and mocking the class in the unit tests. Open to other thoughts!

fluiddot commented on WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/VideoPressBlockProcessor.kt:
```diff
@@ -0,0 +1,198 @@
+            jsonAttributes.apply {
+                addProperty(ID_ATTRIBUTE, Integer.parseInt(mRemoteId))
+                addProperty(GUID_ATTRIBUTE, mRemoteGuid)
+                when (src) {
+                    null -> Unit
+                    else -> jsonAttributes.addProperty(SRC_ATTRIBUTE, mRemoteUrl)
+                }
```
In the [iOS implementation](https://github.com/wordpress-mobile/WordPress-iOS/blob/0e5eac244218a31a90e23fb9ef8fa82967c2bb48/WordPress/Classes/ViewRelated/Gutenberg/Processors/GutenbergVideoPressUploadProcessor.swift#L65-L84) we aren't processing the `src` attribute, so not sure if we need to assign the remote URL here 🤔 .

SiobhyB replied:
In my testing the `src` attribute was being added on Android when I replaced the video. From the results of your testing, it seems there are some wider issues with replacing videos, so I'll do some more digging and reply here when I've figured out what's happening. 🤔 Thanks for your thorough review!

fluiddot replied:
Right, the `src` attribute is being added with the local file URL. This is needed to determine that the upload is in progress. However, I'm not sure if we need to update the value in the processor, I think we can leave it with its original value, at least that's what I did in the iOS implementation. Alternatively, we could remove it when the upload finishes, as [we do in the block logic](https://github.com/Automattic/jetpack/blob/3b4d09a3bc4c4aabd08d532d214e22c0bd7bf17d/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx#L136).

SiobhyB replied:
Ah! I see, thank you for expanding. I agree it'd be best to match the iOS implementation, so I've removed this code in https://github.com/wordpress-mobile/WordPress-Android/pull/18187/commits/f8cd6a8865370c1c131382a23c49135d4a95682b. 🙇‍♀️

fluiddot commented on WordPress/src/test/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/TestContent.kt:
```diff
@@ -659,6 +660,22 @@ $newRefactoredGalleryBlockInnerBlocks</figure>
+<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">https://videopress.com/v/${videoPressGuid}?resizeToParent=true&amp;cover=true&amp;preloadContent=metadata&amp;useAverageColor=true</div></figure>
+<!-- /wp:videopress/video -->"""
+
+    const val oldVideoPressBlockWithAttrs = """<!-- wp:videopress/video {"autoplay":true,"controls":false,"description":"","id":${localMediaId},"loop":true,"muted":true,"playsinline":true,"poster":"https://test.files.wordpress.com/2022/02/265-5000x5000-1.jpeg","preload":"none","seekbarColor":"#abb8c3","seekbarLoadingColor":"#cf2e2e","seekbarPlayedColor":"#9b51e0","title":"Demo title","useAverageC ...
+<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"></figure>
+<!-- /wp:videopress/video -->"""
```
I realized that the typical case when a VideoPress block already has attributes is a video replacement. In that case, the HTML also contains the `div` element, e.g.: 
```
<!-- wp:videopress/video {"autoplay":true,"controls":false,"description":"","id":${localMediaId},"loop":true,"muted":true,"playsinline":true,"poster":"https://test.files.wordpress.com/2022/02/265-5000x5000-1.jpeg","preload":"none","seekbarColor":"#abb8c3","seekbarLoadingColor":"#cf2e2e","seekbarPlayedColor":"#9b51e0","title":"Demo title","useAverageColor":false} -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">
https://videopress.com/v/AbCdE?resizeToParent=true&amp;cover=true&amp;autoPlay=true&amp;controls=false&amp;loop=true&amp;muted=true&amp;persistVolume=false&amp;playsinline=true&amp;posterUrl=https%3A%2F%2Ftest.files.wordpress.com%2F2022%2F02%2F265-5000x5000-1.jpeg&amp;preloadContent=none&amp;sbc=%23abb8c3&amp;sbpc=%239b51e0&amp;sblc=%23cf2e2e
</div></figure>
<!-- /wp:videopress/video -->
```

Maybe we should use this HTML for this test case, WDYT?

fluiddot replied:
This is related to the issue I mentioned in [the review](https://github.com/wordpress-mobile/WordPress-Android/pull/18187#pullrequestreview-1365349094).

fluiddot replied:
Solving this thread as I couldn't reproduce the issue again ([reference](https://github.com/wordpress-mobile/WordPress-Android/pull/18187#pullrequestreview-1366545133)).

fluiddot reviewed (commented):
Thanks @SiobhyB for adding the upload processor for the VideoPress block, awesome work 🏅  !

I tested the following cases and ~~found an issue in the latter~~ works as expected:
- Upload video in an empty VideoPress block outside the editor. ✅ 
- Replace video in a VideoPress block outside the editor. ❌ => ✅ 

**UPDATE:** As mentioned in https://github.com/wordpress-mobile/WordPress-Android/pull/18187#pullrequestreview-1366545133, I couldn't reproduce the issue I spotted in the test case _Replace video in a VideoPress block outside the editor._

<details><summary>More information about the no longer reproducible issue</summary>

Seems it's generating unexpected HTML, specifically, it's duplicating the `div` element:

**HTML before replacing the video:**
```
<!-- wp:videopress/video {"id":5110,"guid":"XY81p4ON"} -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">
https://videopress.com/v/XY81p4ON?resizeToParent=true&amp;cover=true&amp;preloadContent=metadata&amp;useAverageColor=true
</div></figure>
<!-- /wp:videopress/video →
```

**Generated HTML after the upload:**
```
<!-- wp:videopress/video {"id":5242,"guid":"vU9iE4t3"} -->
<figure class="wp-block-videopress-video wp-block-jetpack-videopress jetpack-videopress-player"><div class="jetpack-videopress-player__wrapper">
https://videopress.com/v/XY81p4ON?resizeToParent=true&amp;cover=true&amp;preloadContent=metadata&amp;useAverageColor=true
</div><div class="jetpack-videopress-player__wrapper">https://videopress.com/v/vU9iE4t3?resizeToParent=true&amp;cover=true&amp;preloadContent=metadata&amp;useAverageColor=true</div></figure>
<!-- /wp:videopress/video -->
```

<img src=https://user-images.githubusercontent.com/14905380/228887610-473ada1d-387e-4dab-a0cf-beda81094422.jpg width=250>

Let me know if you manage to reproduce it, thanks 🙇 !

**UPDATE:** As a side note, I couldn't reproduce the issue on iOS.

</details>

SiobhyB committed:
refactor: Re-use existing UriWrapper for encoding

SiobhyB commented:
@fluiddot, I haven't been able to replicate the issue from some initial testing. 🤔 I'll keep trying, but wanted to share a recording of the steps I'm going through in case you can spot anything immediately different to the steps you're using to replicate:

https://user-images.githubusercontent.com/2998162/228949323-8608f11b-2499-439e-8a18-2292b08b8f8c.mp4

fluiddot reviewed (commented):
> @fluiddot, I haven't been able to replicate the issue from some initial testing. 🤔 I'll keep trying, but wanted to share a recording of the steps I'm going through in case you can spot anything immediately different to the steps you're using to replicate:

Oh, interesting. I've just tested again and I can't reproduce the issue again. Maybe I was testing using wrong branches of Jetpack/Gutenberg, sorry for the false alarm. I confirm the upload processor is working as expected 🎊 .

I'll hold my approval until we address the issue on the unit tests ([reference](https://github.com/wordpress-mobile/WordPress-Android/pull/18187#discussion_r1153623085)), but the PR as is LGTM ✅.

SiobhyB committed:
fix: Leave src unchanged following re-upload

This commit follows the feedback here: https://github.com/wordpress-mobile/WordPress-Android/pull/18187#discussion_r1153418105

SiobhyB committed:
refactor: Remove unused src attributes

SiobhyB committed:
Merge branch 'trunk' into gutenberg/videopress-completion-processor

SiobhyB commented:
> Oh, interesting. I've just tested again and I can't reproduce the issue again. Maybe I was testing using wrong branches of Jetpack/Gutenberg, sorry for the false alarm. I confirm the upload processor is working as expected 🎊 .

Thanks for double-checking @fluiddot! Glad to know that this worked for you with the latest branches 🙇‍♀️ 

> I'll hold my approval until we address the issue on the unit tests (https://github.com/wordpress-mobile/WordPress-Android/pull/18187#discussion_r1153623085), but the PR as is LGTM ✅.

I've gone ahead to address the issues as well as the other points in your review. Thanks again!

SiobhyB committed:
Empty commit to retrigger checks.

fluiddot reviewed (approved):
I've tested the last changes and worked as expected, LGTM 🎊 !

⚠️ Before merging, it would be great to have a second review from an Android developer to verify the code. I already covered the functionality, so it would be only needed to check that the code from a native POV is ok.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/VideoPressBlockProcessor.kt:
```diff
@@ -0,0 +1,193 @@
+
+        if (videoPressElements != null) {
+            val url = getVideoPressURL(mRemoteGuid)
+            videoPressElements.append("<div class='jetpack-videopress-player__wrapper'>$url</div>")
```
Just to confirm: it's expected to have two `_` between `player` and `wrapper`, right? Like: `player__wrapper`

SiobhyB replied:
That's right, it looks a little off at first glance, but matches the HTML generated by the VideoPress block and is inline with [the BEM naming convention](https://getbem.com/naming/) for CSS. Appreciate you double-checking!

RenanLukas reviewed (approved):
Thanks for the PR, @SiobhyB . It's working as expected.

I've left a small question but I'm approving because it's working and code LGTM :shipit:

SiobhyB commented:
@RenanLukas, thanks so much for reviewing! 🙌

mkevins commented:
Thank you for adding processors support for VideoPress block! :tada:
